docs/designs/ps_design.md

```diff
@@ -1,53 +1,56 @@
 
-## PS
+In conclusion, a high preformance distributed parameter server architecture with scalability and fault-tolerance is needed.
 
-### Compoments
+Following is the overview architecture diagram:
 
-![pserver](../images/pserver.png)
+![parameter server](../images/parameter_server.png)
 
-PS contains two main compoments:
+We will answer following questions in next sections:
 
-- KVStore
-- Optimizer
+- How to make model sharding in parameter server pods?
+- What kind of data structure is used to keep parameters in the memory of parameter server pods?
+- How parameters are initialized in parameter server pods?
+- How parameters are updated in parameter server pods?
+- How to make the parameter server architecture fault-tolerant?
 
-The worker initializes a model, and pushes parameters to KVStore. Before each step of training, the worker pulls the latest model from KVStore. After a round of forward/backward computation, the worker pushes gradients to the PS waiting for processing. Then, optimizer of PS will look up the corresponding parameter from KVStore. At last, it applies gradients to parameters, and updates parameter back to KVStore.
+## Model Sharding
 
+We will distribute model parameters into multiple PS pods, which is called model sharding. There is a hash function that maps a parameter to a PS pod id. And this parameter will be stored at this corresponding PS pod.
 
-### KVStore
+There are several kinds of parameter to be handled separately:
 
-The KVStore could be like following:
+- Very big embedding table: Embedding table is a collection of <item id, embedding vector> pairs. The embedding table name combining an item id becomes a key of the hash function.
+- Dense tensor: The dense tensor parameter name is the key of the hash function.
 
-```python
-class KVStore(object):
-    def __init__(self):
-        self.variable_db = {}
-        self.embedding_table_db = {}
+Assuming that there are N parameter server pods in a training job. We could use a simple round-robin policy at first.
 
-    def get_parameter(self, name):
-        pass
+*PS(i) = hash(key) mod N*
 
-    def get_embedding_vector(self, name, indices):
-        pass
+Each parameter server pod will store some embedding vectors and dense tensor parameter, it only holds a subset of the whole model.
 
-    def set_parameter(self, name, value):
-        pass
+The number of parameter server pods will keep unchanged during a job. We set the priority of parameter server pods higher than worker pods. Thus, worker pods of a lower priority job will be killed first to satisfy another job. If all worker pods of a job are dead , this job is actually stopped. We will restart the job again when the hardware resources are sufficient.
 
-    def set_embedding_vector(self, name, indices, value):
-        pass
-```
+## Model Data Structure
+
+We use two hashmaps to store embeding table parameters and dense tensor parameters.
```

Comment:
Some places in this doc use `parameter`, and some places use `parameters`. It would be better to keep the singular and plural numbers consistent.

Reply:
Done. Add `the` or `a` before `parameter`.